tpmutil/encoding.go

```diff
@@ -120,81 +55,96 @@ func packWithHeader(ch commandHeader, cmd ...interface{}) ([]byte, error) {
+		tmp.Elem().Set(v)
+		return true, tmp.Interface().(SelfMarshaler).TPMMarshal(buf)
+	}
+
+	return false, nil
+}
+
+func packValue(buf io.Writer, v reflect.Value) error {
+	if canMarshal, err := tryMarshal(buf, v); canMarshal {
+		return err
+	}
+
+	switch v.Kind() {
+	case reflect.Ptr:
+		if v.IsNil() {
+			return fmt.Errorf("cannot pack nil %s", v.Type().String())
 		}
-		v := reflect.ValueOf(e)
-		switch v.Kind() {
-		case reflect.Ptr:
-			if err := packType(buf, reflect.Indirect(v).Interface()); err != nil {
-				return err
-			}
-		case reflect.Struct:
-			// TODO(awly): Currently packType cannot handle non-struct fields that implement SelfMarshaler
-			for i := 0; i < v.NumField(); i++ {
-				if err := packType(buf, v.Field(i).Interface()); err != nil {
-					return err
-				}
-			}
-		case reflect.Slice:
-			switch s := e.(type) {
-			case []byte:
-				switch lengthPrefixSize {
-				case tpm20PrefixSize:
-					if err := binary.Write(buf, binary.BigEndian, uint16(len(s))); err != nil {
-						return err
-					}
-				case tpm12PrefixSize:
-					if err := binary.Write(buf, binary.BigEndian, uint32(len(s))); err != nil {
-						return err
-					}
-				default:
-					return fmt.Errorf("lengthPrefixSize is %d, must be either 2 or 4", lengthPrefixSize)
-				}
-				if err := binary.Write(buf, binary.BigEndian, s); err != nil {
-					return err
-				}
-			case RawBytes:
-				if err := binary.Write(buf, binary.BigEndian, s); err != nil {
-					return err
-				}
-			default:
-				return fmt.Errorf("only []byte and RawBytes slices are supported, got %T", e)
-			}
-		default:
-			if err := binary.Write(buf, binary.BigEndian, e); err != nil {
+		return packValue(buf, v.Elem())
+	case reflect.Struct:
+		for i := 0; i < v.NumField(); i++ {
+			f := v.Field(i)
+			if err := packValue(buf, f); err != nil {
 				return err
 			}
 		}
+	default:
+		return binary.Write(buf, binary.BigEndian, v.Interface())
+	}
+	return nil
+}
 
+func packType(buf io.Writer, elts ...interface{}) error {
+	for _, e := range elts {
+		if err := packValue(buf, reflect.ValueOf(e)); err != nil {
+			return err
+		}
 	}
 
 	return nil
 }
 
+// tryUnmarshal attempts to use TPMUnmarshal() to perform the
+// unpack, if the given value implements SelfMarshaler.
+// True is returned if v implements SelfMarshaler & TPMUnmarshal
+// was called, along with an error returned from TPMUnmarshal.
+func tryUnmarshal(buf io.Reader, v reflect.Value) (bool, error) {
+	t := v.Type()
+	if t.Implements(selfMarshalerType) {
+		return true, v.Interface().(SelfMarshaler).TPMUnmarshal(buf)
+	}
+
+	// We might have a non-pointer struct field, which is settable,
+	// but we dont have a pointer with which to implement the interface.
+	// If the pointer of the type implements the interface, and the
+	// value is settable, we should be able to construct a value to call
+	// TPMUnmarshal() with before replacing the value.
+	if v.CanSet() && reflect.PtrTo(t).Implements(selfMarshalerType) {
```

Comment:
Use v.Addr instead, in case unmarshaler depends on some state in v.
Something like:
```
t := v.Type()
if reflect.PtrTo(t).Implements(selfMarshalerType) && v.CanAddr() {
  v = v.Addr()
  t = v.Type()
}

if t.Implements(selfMarshalerType) {
  return true, v.Interface().(SelfMarshaler).TPMUnmarshal(buf)
}
return false, nil
```

Reply:
Addressable is a subset of Settable, Using .Addr() causes a lot of the tests to break.

Reply:
See my example in the other comment thread: `TPMUnmarshal` may depend on v's current state.
Why are the tests failing? I think all the values passed in should be addressable?